app/domain/authentication/authn_jwt/jwt_fetch_signing_key_factory.rb

```diff
@@ -0,0 +1,65 @@
+module Authentication
+  module AuthnJwt
+    # Factory that returns the interface implementation of FetchSigningKey
+    class JwtFetchSigningKeyFactory
+      FETCH_SIGNING_KEY_METHODS = {
+        "provider-uri" => JwtFetchProviderUriSigningKey,
+        "jwks-uri" => JwtFetchJwksUriSigningKey
+      }
+
+      def signin_key_implementation(authenticator_input)
+        @authenticator_input = authenticator_input
+
+        select_fetch_key_method
+        FETCH_SIGNING_KEY_METHODS[@selected_fetch_key_method].new || raise("Fetch key method #{@selected_fetch_key_method} does not exist.")
+      end
+
+      private
+
+      def select_fetch_key_method
+        validate_key_configuration
+
+        if provider_uri_resource_exists?
+          @selected_fetch_key_method = PROVIDER_URI_RESOURCE_NAME
+        elsif jwks_uri_resource_exists?
+          @selected_fetch_key_method = JWKS_URI_RESOURCE_NAME
+        end
+      end
+
+      def validate_key_configuration
+        if (provider_uri_resource_exists? and jwks_uri_resource_exists?) or
+          (!provider_uri_resource_exists? and !jwks_uri_resource_exists?)
+          raise Errors::Authentication::AuthnJwt::InvalidIssuerConfiguration.new(
+            PROVIDER_URI_RESOURCE_NAME,
+            JWKS_URI_RESOURCE_NAME
+          )
+        end
+      end
+
+      def provider_uri_resource_exists?
+        !provider_uri_resource.nil?
+      end
+
+      def jwks_uri_resource_exists?
+        !jwks_uri_resource.nil?
+      end
+
+      def provider_uri_resource
+        @provider_uri_resource ||= resource(PROVIDER_URI_RESOURCE_NAME)
+      end
+
+      def jwks_uri_resource
+        @jwks_uri_resource ||= resource(JWKS_URI_RESOURCE_NAME)
+      end
+
+      def resource(resource_name)
+        ::Resource[resource_id(resource_name)]
```

Comment:
should `Resource` be a dependency?

Reply:
The factory is not a command_class so Im not sure about what you mean by making `Resource` a dependency

Reply:
can you add a UT for the factory file? once you'll do that you will see that you need to mock the `Resource` class (as you don't want to really go to the DB).

How can we mock this in a factory class? we can change the signature of the `create` method (which is now called `signin_key_implementation`) from:
```
def signin_key_implementation(authenticator_input)
```
to:
```
def signin_key_implementation(authenticator_input, resource_class: ::Resource)
```

this way your app code that uses this class doesn't need to be bothered by this object and you can mock it in a UT.

Reply:
I get it now, thanks